File: apps/frontend/src/types/events.ts
/** SSE event types — matches backend schemas/events.py */

export interface PhaseEvent {
  v: number;
  phase: string;
  turn_index: number;
  at: string;
}

export interface TokenEvent {
  v: number;
  delta: string;
  turn_index: number;
}

export interface AnalysisSnapshotEvent {
  v: number;
  turn_index: number;
  lead_status: string;
  lead_score: number | null;
  lead_score_delta: number | null;
  next_score_contributor: string | null;
  industry: unknown;
  business_size: unknown;
  pain_points: unknown[];
  recommended_services: unknown[];
  conversation_progress: unknown;
  qualification_status: unknown;
}

export interface ErrorEvent {
  v: number;
  code: string;
  message: string;
  retryable: boolean;
  turn_index: number;
}

export interface DoneEvent {
  v: number;
  turn_index: number;
  client_turn_id?: string;
  message_id?: string;
  finish_reason: string;
  consultation_complete: boolean;
  consultation_id?: string;
}

export type SSEEvent =
  | { type: "phase"; data: PhaseEvent }
  | { type: "token"; data: TokenEvent }
  | { type: "analysis_snapshot"; data: AnalysisSnapshotEvent }
  | { type: "error"; data: ErrorEvent }
  | { type: "done"; data: DoneEvent };

File: apps/backend/app/schemas/events.py
"""SSE event schemas — the streaming transport contract.

Each event type maps to a Pydantic model. Events are serialised as
``event: <type>\\ndata: <json>`` lines in the SSE stream.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class PhaseEvent(BaseModel):
    """Emitted when a pipeline phase starts."""

    v: int = 1
    phase: str
    turn_index: int
    at: str


class TokenEvent(BaseModel):
    """Emitted for each chunk of response text."""

    v: int = 1
    delta: str
    turn_index: int


class ErrorEvent(BaseModel):
    """Emitted on a recoverable turn failure."""

    v: int = 1
    code: str
    message: str
    retryable: bool = True
    turn_index: int


class DoneEvent(BaseModel):
    """Always the final event in a stream."""

    v: int = 1
    turn_index: int
    client_turn_id: str | None = None
    message_id: str | None = None
    finish_reason: str = "complete"
    consultation_complete: bool = False
    consultation_id: str | None = None
